### PR TITLE
Azure: use per-vmss vmssvm incremental cache

### LINKED ISSUE #93106: Azure cloud provider hits the VMSS VMs List API more than it should
**What happened**:

Running a non-AKS k8s v1.18.5  cluster on Azure, with 50 VMSS attached (each containing one or two nodes).

When we do attach several new VMSSes, or create/replace several nodes, the controller-managed gets quickly throttled by Azure API, for instance like so:
```
kube-controller-manager[2014310]: E0708 08:33:13.752989 2014310 azure_vmss.go:570]
VirtualMachineScaleSetsClient.List failed:
&{true 0 2020-07-08 08:34:43.665286837 +0000 UTC m=+1598.271278798
azure cloud provider throttled for operation VMSSList with reason "client throttled"}
```

Having a closer look at the cloud provider code, we see controller-manager would re-scan all attached VMSS on cache miss or expired TTL. Not just the one that needs a refresh.

**What you expected to happen**:

Number of VMSS List calls should be linear (eg. one for a new instance's showing up or being deleted), and not O(number of vmss).

**How to reproduce it (as minimally and precisely as possible)**:

Attach many VMSS (even if they contain a single VM) to a cluster, then attach more or create many pods with remote PVs.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.5
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): Ubuntu 20.04
- Kernel (e.g. `uname -a`): 5.4.0
- Controller-manager cloud-config:
```
{
  "subscriptionId": "xxx",
  "resourceGroup": "yyy",
  "location": "westus2",
  "vnetName": "www",
  "vnetResourceGroup": "zzz",
  "vmType": "vmss",
  "loadBalancerSku": "standard",
  "useManagedIdentityExtension": true,
  "excludeMasterFromStandardLB": true,
  "useInstanceMetadata": true,
  "cloudConfigType": "file",
  "vmssCacheTTLInSeconds": 900,
  "vmssVirtualMachinesCacheTTLInSeconds": 900,
  "vmCacheTTLInSeconds": 900
}
```

/kind bug
/sig cloud-provider
/area provider/azure

### PR DESCRIPTION
**Pull request by bpineau**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Azure's cloud provider VMSS VMs API accesses are mediated through a cache holding and refreshing all VMSS together.

Due to that we hit VMSSVM.List API more often than we could: an instance's cache miss or expiration should only require a single VMSS re-list, while it's currently O(n) relative to the number of attached Scale Sets.

Under hard pressure (clusters with many attached VMSS that can't all be listed in one sequence of successive API calls) the controller manager might be stuck trying to re-list everything from scratch, then aborting the whole operation due to rate limits, affecting the whole Subscription.

This patch replaces the global VMSS VMs cache by per-VMSS VMs caches.  Refreshes (VMSS VMs lists) are scoped to the single relevant VMSS; under severe throttling the various caches can be incrementally refreshed.

**Which issue(s) this PR fixes**:

Fixes #93106

**Special notes for your reviewer**:

We are assuming VMSS nodes are named from VMSS' computerNamePrefix+id (or vmssName+id, when computerNamePrefix isn't specified), as described [https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-instance-ids](here) and [https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/2018-10-01/virtualmachinescalesets#virtualmachinescalesetosprofile-object](there). Are there special cases not covered by that doc? if so we can probably complement that optimistic lookup (trying the vmss matching name prefix first, happy path) by a fallback to a scan over remaining scale sets. The non-optimal fallback path would still be an improvement as (in case of throttling) we keep partial results (per VMSS caches), refreshes are incremental.

**Does this PR introduce a user-facing change?**:

```release-note
Azure: per VMSS VMSS VMs cache to prevent throttling on clusters having many attached VMSS
```

/assign @andyzhangx @feiskyer
/sig cloud-provider
/area provider/azure

### CODE REVIEW

**Comment by andyzhangx** [comment]
/ok-to-test
/priority important-soon
/sig cloud-provider
/area provider/azure
thanks for the contribution @bpineau

**Comment by liggitt** [comment]
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93107/pull-kubernetes-node-e2e/1283340355466432512 flake is #93210

**Comment by nilo19** [comment]
```
1/37 Tests Failed. | expand_less
-- | --
staticcheck expand_less6m57sErrors from staticcheck: vendor/k8s.io/legacy-cloud-providers/azure/azure_vmss_cache_test.go:118:2: this value of err is never used (SA4006)  Please review the above warnings. You can test via:   hack/verify-staticcheck.sh <failing package> If the above warnings do not make sense, you can exempt the line or file. See:   https://staticcheck.io/docs/#ignoring-problems | staticcheck expand_less | 6m57s | Errors from staticcheck: vendor/k8s.io/legacy-cloud-providers/azure/azure_vmss_cache_test.go:118:2: this value of err is never used (SA4006)  Please review the above warnings. You can test via:   hack/verify-staticcheck.sh <failing package> If the above warnings do not make sense, you can exempt the line or file. See:   https://staticcheck.io/docs/#ignoring-problems
staticcheck expand_less | 6m57s
Errors from staticcheck: vendor/k8s.io/legacy-cloud-providers/azure/azure_vmss_cache_test.go:118:2: this value of err is never used (SA4006)  Please review the above warnings. You can test via:   hack/verify-staticcheck.sh <failing package> If the above warnings do not make sense, you can exempt the line or file. See:   https://staticcheck.io/docs/#ignoring-problems
```
could you please fix the error and re-run the tests?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss.go [hunk 1]
```diff
@@ -590,6 +614,66 @@ func (ss *scaleSet) listScaleSets(resourceGroup string) ([]string, error) {
...
+				node.resourceGroup = v.resourceGroup
+				return false
+			}
+
+			return true
+		})
+		return node, nil
+	}
+
+	if len(nodeName) < 6 {
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss.go:654]: could we use `getScaleSetVMInstanceID()` here to check whether a node is VMSS instance or not?

**Review by feiskyer — COMMENTED** [review]
Thanks for fixing the issue. Add it to v1.19 milestone since it would fix a serious bug for large number of VMSS scenarios.
/milestone v1.19

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve
/retest

**Pull request merged**